RxSwift/Observables/Implementations/SwitchIfEmpty.swift

```diff
@@ -0,0 +1,127 @@
+//
+//  SwitchIfEmpty.swift
+//  Rx
+//
+//  Created by sergdort on 23/12/2016.
+//  Copyright © 2016 Krunoslav Zaher. All rights reserved.
+//
+
+import Foundation
+
+
+final class SwitchIfEmpty<SourceType, S: ObservableConvertibleType>: Producer<S.E> where S.E == SourceType {
+    
+    typealias Selector = (Void) throws -> S
+    
+    fileprivate let _source: Observable<SourceType>
+    fileprivate let _resultSelector: Selector
+    
+    init(source: Observable<SourceType>, resultSelector: @escaping Selector) {
+        _source = source
+        _resultSelector = resultSelector
+    }
+    
+    override func run<O : ObserverType>(_ observer: O, cancel: Cancelable) -> (sink: Disposable, subscription: Disposable) where O.E == S.E {
+        let sink = SwitchIfEmptySink(selector: _resultSelector,
+                                     observer: observer,
+                                     cancel: cancel)
+        let subscription = sink.run(_source)
+        
+        return (sink: sink, subscription: subscription)
+    }
+}
+
+final class SwitchIfEmptySink<SourceType, S: ObservableConvertibleType, O: ObserverType>: Sink<O>
+    , ObserverType
+    , LockOwnerType
+    , SynchronizedOnType where S.E == O.E, SourceType == S.E {
+    typealias E = SourceType
+    typealias Selector = (Void) throws -> S
+    
+    let _lock = NSRecursiveLock()
```

Comment:
We don't need guard this or `SwitchIfEmptySinkIter` with `_lock`, `LockOwnerType`, `SynchronizedOnType`.

Each of the sequence should be already synchronized internally, and second sequence can only start once first one has completed.